core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java

```diff
@@ -306,7 +312,8 @@ private void buildUberJar0(CurateOutcomeBuildItem curateOutcomeBuildItem,
                 final AppArtifact depArtifact = appDep.getArtifact();
 
                 // Exclude files that are not jars (typically, we can have XML files here, see https://github.com/quarkusio/quarkus/issues/2852)
-                if (!isAppDepAJar(depArtifact)) {
+                // And exclude dependencies that should not be kept
+                if (!isAppDepAJar(depArtifact) || !keepDependency(appDep, outputTargetBuildItem.getOptionalDependencies())) {
```

Comment:
You can rename and enhance `isAppDepAJar(...)` to include your filtering logic. It doesn't look like there is a reason to keep two methods.